src/connection.rs

```diff
@@ -63,23 +80,40 @@ impl Connection {
 
             // There is not enough buffered data to read a frame. Attempt to
             // read more data from the socket.
-            //
-            // On success, the number of bytes is returned. `0` indicates "end
-            // of stream".
-            if 0 == self.stream.read_buf(&mut self.buffer).await? {
-                // The remote closed the connection. For this to be a clean
-                // shutdown, there should be no data in the read buffer. If
-                // there is, this means that the peer closed the socket while
-                // sending a frame.
-                if self.buffer.is_empty() {
+            match self.maybe_read_bytes().await? {
+                ConnectionState::Open => continue,
+                ConnectionState::Closed | ConnectionState::Reset => {
+                    if !self.buffer.is_empty() {
+                        warn! {
+                            incomplete =? self.buffer,
+                            "connection closed with incomplete frame"
+                        };
+                    }
                     return Ok(None);
-                } else {
-                    return Err("connection reset by peer".into());
                 }
             }
         }
     }
 
+    /// Attempt to read bytes from the connection.
+    async fn maybe_read_bytes(&mut self) -> io::Result<ConnectionState> {
+        match self.stream.read_buf(&mut self.buffer).await {
+            // the connection was closed gracefully
+            Ok(0) => Ok(ConnectionState::Closed),
+            // the connection is still open
+            Ok(_) => Ok(ConnectionState::Open),
+            // the connection was closed abruptly by the peer
+            Err(e) if e.kind() == ConnectionReset => {
+                warn! {
+                    "connection closed abruptly by peer"
+                };
```

Comment:
nit, take it or leave it: i have basically never seen anyone use curly braces with the function-like tracing macros; generally i would expect
```suggestion
                warn!("connection closed abruptly by peer");
```